newton/_src/geometry/sdf_hydroelastic.py

```diff
@@ -0,0 +1,1925 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any
+
+import warp as wp
+
+from ..sim.model import Model
+from .collision_core import build_pair_key2, sat_box_intersection
+from .contact_data import ContactData
+from .contact_reduction import (
+    NUM_NORMAL_BINS,
+    NUM_SPATIAL_DIRECTIONS,
+    get_slot,
+    get_spatial_direction_2d,
+    project_point_to_plane,
+)
+from .sdf_contact import sample_sdf_extrapolated
+from .sdf_mc import get_mc_tables, mc_calc_face
+from .sdf_utils import SDFData
+from .utils import scan_with_total
+
+vec8f = wp.types.vector(length=8, dtype=wp.float32)
+
+MIN_FRICTION = 1e-4
+EPS_LARGE = 1e-8
+EPS_SMALL = 1e-20
+
+
+@wp.func
+def int_to_vec3f(x: wp.int32, y: wp.int32, z: wp.int32):
+    return wp.vec3f(float(x), float(y), float(z))
+
+
+@dataclass
+class HydroelasticContactSurfaceData:
+    """
+    Data container for hydroelastic contact surface visualization.
+
+    Contains the vertex arrays and metadata needed for rendering
+    the contact surface triangles from hydroelastic collision detection.
+    """
+
+    contact_surface_point: wp.array(dtype=wp.vec3f)
+    """World-space positions of contact surface triangle vertices (3 per face)."""
+    contact_surface_depth: wp.array(dtype=wp.float32)
+    """Penetration depth at each face centroid."""
+    contact_surface_shape_pair: wp.array(dtype=wp.vec2i)
+    """Shape pair indices (shape_a, shape_b) for each face."""
+    face_contact_count: wp.array(dtype=wp.int32)
+    """Array containing the number of face contacts."""
+    max_num_face_contacts: int
+    """Maximum number of face contacts (buffer size)."""
+
+
+@dataclass
+class SDFHydroelasticConfig:
+    """
+    Controls properties of SDF hydroelastic collision handling.
+    """
+
+    reduce_contacts: bool = True
+    """Whether to reduce contacts."""
+    buffer_mult_broad: int = 1
+    """Multiplier for buffer size for broadphase."""
+    buffer_mult_iso: int = 1
+    """Multiplier for buffer size for iso voxel traversal."""
+    buffer_mult_contact: int = 1
+    """Multiplier for buffer size for face contacts."""
+    grid_size: int = 256 * 8 * 128
+    """Grid size for contact handling. Can be tuned for performance."""
+    output_contact_surface: bool = False
+    """Whether to output hydroelastic contact surface vertices for visualization."""
+    betas: tuple[float, float] = (10.0, -0.5)
+    """Penetration beta values."""
+    sticky_contacts: float = 0.0
+    """Stickiness factor for temporal contact persistence."""
+    normal_matching: bool = True
+    """Whether to match the aggregated force direction."""
```

Comment:
I have maybe an idea of what this does from your talk, but this really needs more explanation either here or in a doc section.